### PR TITLE
add modular pcl packages for ubuntu and debian

### PR DESCRIPTION
**Pull request by Aposhian**
Please add the following dependency to the rosdep database.

## Package name:
- libpcl-common
- libpcl-features
- libpcl-filters
- libpcl-io
- libpcl-kdtree
- libpcl-keypoints
- libpcl-ml
- libpcl-octree
- libpcl-outofcore
- libpcl-people
- libpcl-recognition
- libpcl-registration
- libpcl-sample-consensus
- libpcl-search
- libpcl-segmentation
- libpcl-stereo
- libpcl-surface
- libpcl-tracking
- libpcl-visualization

## Package Upstream Source:

https://github.com/PointCloudLibrary/pcl

## Purpose of using this:
Allow Ubuntu and Debian users to only pull in the dependencies they need, since all of PCL is quite large (including GUI dependencies).

Other platforms which don't have these modular packages will simply fallback to their catchall packages, maintaining compatibility

Distro packaging links:

## Links to Distribution Packages

- Debian: https://packages.debian.org/
  - https://packages.debian.org/stable/libpcl-apps1.11
  - https://packages.debian.org/oldstable/libpcl-apps1.9
  - https://packages.debian.org/stable/libpcl-common1.11
  - https://packages.debian.org/oldstable/libpcl-common1.9
  - https://packages.debian.org/stable/libpcl-features1.11
  - https://packages.debian.org/oldstable/libpcl-features1.9
  - https://packages.debian.org/stable/libpcl-filters1.11
  - https://packages.debian.org/oldstable/libpcl-filters1.9
  - https://packages.debian.org/stable/libpcl-io1.11
  - https://packages.debian.org/oldstable/libpcl-io1.9
  - https://packages.debian.org/stable/libpcl-kdtree1.11
  - https://packages.debian.org/oldstable/libpcl-kdtree1.9
  - https://packages.debian.org/stable/libpcl-keypoints1.11
  - https://packages.debian.org/oldstable/libpcl-keypoints1.9
  - https://packages.debian.org/stable/libpcl-ml1.11
  - https://packages.debian.org/oldstable/libpcl-ml1.9
  - https://packages.debian.org/stable/libpcl-octree1.11
  - https://packages.debian.org/oldstable/libpcl-octree1.9
  - https://packages.debian.org/stable/libpcl-outofcore1.11
  - https://packages.debian.org/oldstable/libpcl-outofcore1.9
  - https://packages.debian.org/stable/libpcl-people1.11
  - https://packages.debian.org/oldstable/libpcl-people1.9
  - https://packages.debian.org/stable/libpcl-recognition1.11
  - https://packages.debian.org/oldstable/libpcl-recognition1.9
  - https://packages.debian.org/stable/libpcl-registration1.11
  - https://packages.debian.org/oldstable/libpcl-registration1.9
  - https://packages.debian.org/stable/libpcl-sample-consensus1.11
  - https://packages.debian.org/oldstable/libpcl-sample-consensus1.9
  - https://packages.debian.org/stable/libpcl-search1.11
  - https://packages.debian.org/oldstable/libpcl-search1.9
  - https://packages.debian.org/stable/libpcl-segmentation1.11
  - https://packages.debian.org/oldstable/libpcl-segmentation1.9
  - https://packages.debian.org/stable/libpcl-stereo1.11
  - https://packages.debian.org/oldstable/libpcl-stereo1.9
  - https://packages.debian.org/stable/libpcl-surface1.11
  - https://packages.debian.org/oldstable/libpcl-surface1.9
  - https://packages.debian.org/stable/libpcl-tracking1.11
  - https://packages.debian.org/oldstable/libpcl-tracking1.9
  - https://packages.debian.org/stable/libpcl-visualization1.11
  - https://packages.debian.org/oldstable/libpcl-visualization1.9
- Ubuntu: https://packages.ubuntu.com/
   - https://packages.ubuntu.com/bionic/libpcl-apps1.8
   - https://packages.ubuntu.com/focal/libpcl-apps1.10
   - https://packages.ubuntu.com/jammy/libpcl-apps1.12
   - https://packages.ubuntu.com/bionic/libpcl-common1.8
   - https://packages.ubuntu.com/focal/libpcl-common1.10
   - https://packages.ubuntu.com/jammy/libpcl-common1.12
   - https://packages.ubuntu.com/bionic/libpcl-features1.8
   - https://packages.ubuntu.com/focal/libpcl-features1.10
   - https://packages.ubuntu.com/jammy/libpcl-features1.12
   - https://packages.ubuntu.com/bionic/libpcl-filters1.8
   - https://packages.ubuntu.com/focal/libpcl-filters1.10
   - https://packages.ubuntu.com/jammy/libpcl-filters1.12
   - https://packages.ubuntu.com/bionic/libpcl-io1.8
   - https://packages.ubuntu.com/focal/libpcl-io1.10
   - https://packages.ubuntu.com/jammy/libpcl-io1.12
   - https://packages.ubuntu.com/bionic/libpcl-kdtree1.8
   - https://packages.ubuntu.com/focal/libpcl-kdtree1.10
   - https://packages.ubuntu.com/jammy/libpcl-kdtree1.12
   - https://packages.ubuntu.com/bionic/libpcl-keypoints1.8
   - https://packages.ubuntu.com/focal/libpcl-keypoints1.10
   - https://packages.ubuntu.com/jammy/libpcl-keypoints1.12
   - https://packages.ubuntu.com/bionic/libpcl-ml1.8
   - https://packages.ubuntu.com/focal/libpcl-ml1.10
   - https://packages.ubuntu.com/jammy/libpcl-ml1.12
   - https://packages.ubuntu.com/bionic/libpcl-octree1.8
   - https://packages.ubuntu.com/focal/libpcl-octree1.10
   - https://packages.ubuntu.com/jammy/libpcl-octree1.12
   - https://packages.ubuntu.com/bionic/libpcl-outofcore1.8
   - https://packages.ubuntu.com/focal/libpcl-outofcore1.10
   - https://packages.ubuntu.com/jammy/libpcl-outofcore1.12
   - https://packages.ubuntu.com/bionic/libpcl-people1.8
   - https://packages.ubuntu.com/focal/libpcl-people1.10
   - https://packages.ubuntu.com/jammy/libpcl-people1.12
   - https://packages.ubuntu.com/bionic/libpcl-recognition1.8
   - https://packages.ubuntu.com/focal/libpcl-recognition1.10
   - https://packages.ubuntu.com/jammy/libpcl-recognition1.12
   - https://packages.ubuntu.com/bionic/libpcl-registration1.8
   - https://packages.ubuntu.com/focal/libpcl-registration1.10
   - https://packages.ubuntu.com/jammy/libpcl-registration1.12
   - https://packages.ubuntu.com/bionic/libpcl-sample-consensus1.8
   - https://packages.ubuntu.com/focal/libpcl-sample-consensus1.10
   - https://packages.ubuntu.com/jammy/libpcl-sample-consensus1.12
   - https://packages.ubuntu.com/bionic/libpcl-search1.8
   - https://packages.ubuntu.com/focal/libpcl-search1.10
   - https://packages.ubuntu.com/jammy/libpcl-search1.12
   - https://packages.ubuntu.com/bionic/libpcl-segmentation1.8
   - https://packages.ubuntu.com/focal/libpcl-segmentation1.10
   - https://packages.ubuntu.com/jammy/libpcl-segmentation1.12
   - https://packages.ubuntu.com/bionic/libpcl-stereo1.8
   - https://packages.ubuntu.com/focal/libpcl-stereo1.10
   - https://packages.ubuntu.com/jammy/libpcl-stereo1.12
   - https://packages.ubuntu.com/bionic/libpcl-surface1.8
   - https://packages.ubuntu.com/focal/libpcl-surface1.10
   - https://packages.ubuntu.com/jammy/libpcl-surface1.12
   - https://packages.ubuntu.com/bionic/libpcl-tracking1.8
   - https://packages.ubuntu.com/focal/libpcl-tracking1.10
   - https://packages.ubuntu.com/jammy/libpcl-tracking1.12
   - https://packages.ubuntu.com/bionic/libpcl-visualization1.8
   - https://packages.ubuntu.com/focal/libpcl-visualization1.10
   - https://packages.ubuntu.com/jammy/libpcl-visualization1.12
- Fedora: https://packages.fedoraproject.org/
  - https://packages.fedoraproject.org/pkgs/pcl/pcl/
- Arch: https://www.archlinux.org/packages/
  - not available (only in https://aur.archlinux.org/packages/pcl)
- Gentoo: https://packages.gentoo.org/
  - https://packages.gentoo.org/packages/sci-libs/pcl
- macOS: https://formulae.brew.sh/
  - https://formulae.brew.sh/formula/pcl
- Alpine: https://pkgs.alpinelinux.org/packages
  - https://pkgs.alpinelinux.org/package/edge/testing/x86_64/pcl-libs
- NixOS/nixpkgs: https://search.nixos.org/packages
  - https://search.nixos.org/packages?show=pcl
- openSUSE: https://software.opensuse.org/package/
  - not available (only in https://build.opensuse.org/package/show/science/pcl)

### CODE REVIEW

**Comment by methylDragon** [comment]
Nosetests are failing

**Review by wjwwood — DISMISSED** [review]
lgtm, needs a second review

**Comment by ivanpauno** [comment]
@Aposhian please provide links for each package for each distro, so they can be checked

File: rosdep/base.yaml
Comment (ivanpauno) [inline rosdep/base.yaml:4630]:
```suggestion
  arch: [pcl]
  debian:
```

Comment (ivanpauno) [reply rosdep/base.yaml:4630]: please apply to all the other libs this change

File: rosdep/base.yaml
Comment (ivanpauno) [inline rosdep/base.yaml:4639]: same:
```suggestion
  rhel: [pcl]
```

File: rosdep/base.yaml
Comment (ivanpauno) [inline rosdep/base.yaml:4638]:
```suggestion
  openembedded: [pcl@meta-ros-common]
  opensuse: [pcl]
```

**Review by ivanpauno — CHANGES_REQUESTED** [review]
Please apply the suggestions to all the other packages

**Comment by ivanpauno** [comment]
> May be a dup of https://github.com/ros/rosdistro/pull/35137, I will double check when reviewing again

It's not exactly a dup, but there's some duplication.
I will merge the first one that's ready, the other one will need a rebase after that.

File: rosdep/base.yaml
Comment (mvieth) [reply rosdep/base.yaml:4630]: In my pull request, a check complained that the pcl package could not be found on arch and opensuse, see here: https://github.com/ros/rosdistro/pull/35137#issuecomment-1309476280

File: rosdep/base.yaml
Comment (mvieth) [inline rosdep/base.yaml:4629]: You could add `alpine: [pcl-libs]` as well, and also the packages on homebrew like here: https://github.com/ros/rosdistro/pull/35137/files#diff-86e444436e8b275218b0f9fce40452b3a5a24dad5f9bc554b8ec9737134a3684

**Review by mvieth — COMMENTED** [review]
Hi, I am the author of the other pull request (https://github.com/ros/rosdistro/pull/35137) which follows a similar idea. I am willing to close my pull request in favour of this one if you make the suggested changes

File: rosdep/base.yaml
Comment (ivanpauno) [reply rosdep/base.yaml:4630]: Thanks for the info.
I was looking at the preexisting `libpcl-all-dev` rule.
We should fix that one later, but no need to do that here.

File: rosdep/base.yaml
Comment (ivanpauno) [reply rosdep/base.yaml:4638]: See https://github.com/ros/rosdistro/pull/35289#discussion_r1025180890, resolving

**Comment by Aposhian** [comment]
Ok. Thanks for pointing out the sibling PR @ivanpauno. Best I can tell the differences are:
- this PR adds some additional packages
- this PR does not yet add macOS config
- some inconsistencies with how arch, alpine, and opensuse is handled

I'm happy to resolve the requested tasks, but it may take me a few days.

Happy to see there are others who don't want to pull in all of PCL at a time!

**Comment by Aposhian** [comment]
Ok I have:
- added homebrew config from the other PR
- added `alpine: [pcl-libs]` for everything.

I think the last outstanding thing is to add package links to the PR description?

**Comment by Aposhian** [comment]
@ivanpauno @mvieth I believe I have made all requested changes. Let me know if there was anything I missed.

File: rosdep/osx-homebrew.yaml
Comment (mvieth) [inline rosdep/osx-homebrew.yaml:318]: Would be good if you also add the rules for the other packages here: libpcl-apps, libpcl-keypoints, libpcl-ml, libpcl-octree, libpcl-outofcore, libpcl-people, libpcl-recognition, libpcl-registration, libpcl-sample-consensus, libpcl-search, libpcl-stereo, libpcl-tracking, libpcl-visualization

File: rosdep/base.yaml
Comment (mvieth) [inline rosdep/base.yaml:4897]: What about bionic for libpcl-segmentation, libpcl-stereo, libpcl-surface, libpcl-tracking, libpcl-visualization?

File: rosdep/osx-homebrew.yaml
Comment (Aposhian) [reply rosdep/osx-homebrew.yaml:320]: Done

File: rosdep/osx-homebrew.yaml
Comment (ivanpauno) [inline rosdep/osx-homebrew.yaml:226]: what did happen here?
seems that you should restore those lines.



CI output:

> 225:3 [key-duplicates] duplication of key "osx" in mapping

File: rosdep/osx-homebrew.yaml
Comment (ivanpauno) [inline rosdep/osx-homebrew.yaml:343]: alpha order:

```suggestion
libpcl-kdtree:
  osx:
    homebrew:
      packages: [pcl]
libpcl-keypoints:
  osx:
    homebrew:
      packages: [pcl]
```

**Review by ivanpauno — COMMENTED** [review]
A few more changes to go, otherwise lgtm

File: rosdep/osx-homebrew.yaml
Comment (Aposhian) [reply rosdep/osx-homebrew.yaml:226]: Yes that looks like a typo. Will fix.

**Review by ivanpauno — APPROVED** [review]
LGTM!

**Comment by ivanpauno** [comment]
There has been many changes from the first review, so I will wait for a second review to be done.
Thanks @Aposhian @mvieth for the PR!!

**Review by tfoote — APPROVED** [review]

**Comment by Aposhian** [comment]
It looks like we have two approvals. Can we get a merge on this and close it out?

**Pull request merged**
